# Введение в обработку текста на естественном языке

__Автор задач: Блохин Н.В. (NVBlokhin@fa.ru)__

Материалы:
* Макрушин С.В. Лекция "Введение в обработку текста на естественном языке"
* https://www.nltk.org/api/nltk.metrics.distance.html
* https://pymorphy2.readthedocs.io/en/stable/user/guide.html
* https://realpython.com/nltk-nlp-python/
* https://scikit-learn.org/stable/modules/feature_extraction.html

In [1]:
import pymorphy2, random, pandas as pd
import nltk
from nltk import word_tokenize, sent_tokenize
from nltk.stem import SnowballStemmer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.feature_extraction.text import CountVectorizer
from collections import Counter

## Задачи для совместного разбора

In [43]:
!pip install pymorphy2

     -------------------------------------- 55.5/55.5 kB 481.3 kB/s eta 0:00:00
     ---------------------------------------- 8.2/8.2 MB 1.3 MB/s eta 0:00:00
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
  Created wheel for docopt: filename=docopt-0.6.2-py2.py3-none-any.whl size=13724 sha256=77047f4261009601722176a7064a25a223d880ebaa037bfd80720ac3f5c2bb30
  Stored in directory: c:\users\admin\appdata\local\pip\cache\wheels\70\4a\46\1309fc853b8d395e60bafaf1b6df7845bdd82c95fd59dd8d2b
Successfully built docopt


In [1]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping tokenizers\punkt.zip.


True

1. Считайте слова из файла `litw-win.txt` и запишите их в список `words`. При помощи расстояния Левенштейна иправьте опечатку в слове "велечайшим".

In [ ]:
from nltk import edit_distance

2. Разбейте текст из формулировки второго задания на слова. Проведите стемминг и лемматизацию слов.

In [ ]:
text = '''Разбейте текст из формулировки второго задания на слова. Проведите стемминг и лемматизацию слов.'''

3. Преобразуйте предложения из формулировки задания 2 в векторы при помощи `CountVectorizer`. Выведите на экран словарь обученного токенизатора.

In [ ]:
text = '''Разбейте текст из формулировки второго задания на слова. Проведите стемминг и лемматизацию слов.'''

## Лабораторная работа 7

1\. Загрузите данные из файла `ru_recipes_sample.csv` в виде `pd.DataFrame` `recipes` Используя регулярные выражения, удалите из описаний (столбец `description`) все символы, кроме русских букв, цифр и пробелов. Приведите все слова в описании к нижнему регистру. Сохраните полученный результат в столбец `description`.

In [2]:
recipes = pd.DataFrame(pd.read_csv('07_nlp_data/ru_recipes_sample.csv'))
recipes

url  \
0     https://www.povarenok.ru/recipes/show/164365/   
1       https://www.povarenok.ru/recipes/show/1306/   
2      https://www.povarenok.ru/recipes/show/10625/   
3     https://www.povarenok.ru/recipes/show/167337/   
4      https://www.povarenok.ru/recipes/show/91919/   
...                                             ...   
3462   https://www.povarenok.ru/recipes/show/54574/   
3463  https://www.povarenok.ru/recipes/show/113494/   
3464   https://www.povarenok.ru/recipes/show/83228/   
3465  https://www.povarenok.ru/recipes/show/172238/   
3466   https://www.povarenok.ru/recipes/show/52794/   

                                                   name  \
0                    Густой молочно-клубничный коктейль   
1                                              Рулетики   
2                                   Салат "Баклажанчик"   
3         Куриные котлеты с картофельным пюре в духовке   
4                               Рецепт вишневой наливки   
...                                                 ...   
3462                                               Мшош   
3463                   Мясные треугольники с баклажаном   
3464                               "Болоньез" по-новому   
3465               Варенье из одуванчиков с апельсинами   
3466  Три корочки хлеба под соусом болоньез для Пино...   

                                            ingredients  \
0     {'Молоко': '250 мл', 'Клубника': '200 г', 'Сах...   
1     {'Сыр твердый': None, 'Чеснок': None, 'Яйцо ку...   
2     {'Баклажан': '3 шт', 'Лук репчатый': '2 шт', '...   
3     {'Фарш куриный': '800 г', 'Пюре картофельное':...   
4     {'Вишня': '1 кг', 'Водка': '1 л', 'Сахар': '30...   
...                                                 ...   
3462  {'Чечевица': '1 стак.', 'Лук репчатый': '2 шт'...   
3463  {'Фарш мясной': '400 г', 'Баклажан': '1 шт', '...   
3464  {'Фарш мясной': '400 г', 'Томаты в собственном...   
3465  {'Цветки': '400 г', 'Сахар': '1300 г', 'Апельс...   
3466  {'Хлеб': '3 шт', 'Фарш мясной': '200-250 г', '...   

                                            description  
0     Этот коктейль готовлю из замороженной клубники...  
1                                      Быстро и вкусно.  
2           Сытный, овощной салатик, пальчики оближете.  
3     Картофельное пюре и куриные котлеты - вкусная ...  
4     Вишневая наливка имеет яркий вишневый вкус, ко...  
...                                                 ...  
3462  Для тех, кто любит чечевицу... Вам сюда! Очень...  
3463  Баклажановые фантазии продолжаются! Предлагаю ...  
3464  Мое любимое блюдо лазанья. Но кушать только фа...  
3465  Прошлым летом варила варенье из одуванчиков по...  
3466  - И три корочки хлеба! - сделал заказ Буратино...  

[3467 rows x 4 columns]

In [3]:
recipes['description'].replace(r'[^А-Яа-яЁё0-9\s]', '', regex=True, inplace=True)
recipes['description'] = [value.lower() for value in recipes['description']]
recipes

url  \
0     https://www.povarenok.ru/recipes/show/164365/   
1       https://www.povarenok.ru/recipes/show/1306/   
2      https://www.povarenok.ru/recipes/show/10625/   
3     https://www.povarenok.ru/recipes/show/167337/   
4      https://www.povarenok.ru/recipes/show/91919/   
...                                             ...   
3462   https://www.povarenok.ru/recipes/show/54574/   
3463  https://www.povarenok.ru/recipes/show/113494/   
3464   https://www.povarenok.ru/recipes/show/83228/   
3465  https://www.povarenok.ru/recipes/show/172238/   
3466   https://www.povarenok.ru/recipes/show/52794/   

                                                   name  \
0                    Густой молочно-клубничный коктейль   
1                                              Рулетики   
2                                   Салат "Баклажанчик"   
3         Куриные котлеты с картофельным пюре в духовке   
4                               Рецепт вишневой наливки   
...                                                 ...   
3462                                               Мшош   
3463                   Мясные треугольники с баклажаном   
3464                               "Болоньез" по-новому   
3465               Варенье из одуванчиков с апельсинами   
3466  Три корочки хлеба под соусом болоньез для Пино...   

                                            ingredients  \
0     {'Молоко': '250 мл', 'Клубника': '200 г', 'Сах...   
1     {'Сыр твердый': None, 'Чеснок': None, 'Яйцо ку...   
2     {'Баклажан': '3 шт', 'Лук репчатый': '2 шт', '...   
3     {'Фарш куриный': '800 г', 'Пюре картофельное':...   
4     {'Вишня': '1 кг', 'Водка': '1 л', 'Сахар': '30...   
...                                                 ...   
3462  {'Чечевица': '1 стак.', 'Лук репчатый': '2 шт'...   
3463  {'Фарш мясной': '400 г', 'Баклажан': '1 шт', '...   
3464  {'Фарш мясной': '400 г', 'Томаты в собственном...   
3465  {'Цветки': '400 г', 'Сахар': '1300 г', 'Апельс...   
3466  {'Хлеб': '3 шт', 'Фарш мясной': '200-250 г', '...   

                                            description  
0     этот коктейль готовлю из замороженной клубники...  
1                                       быстро и вкусно  
2              сытный овощной салатик пальчики оближете  
3     картофельное пюре и куриные котлеты  вкусная к...  
4     вишневая наливка имеет яркий вишневый вкус кот...  
...                                                 ...  
3462  для тех кто любит чечевицу вам сюда очень вкус...  
3463  баклажановые фантазии продолжаются предлагаю в...  
3464  мое любимое блюдо лазанья но кушать только фар...  
3465  прошлым летом варила варенье из одуванчиков по...  
3466   и три корочки хлеба  сделал заказ буратино в ...  

[3467 rows x 4 columns]

In [4]:
recipes['description']

0       этот коктейль готовлю из замороженной клубники...
1                                         быстро и вкусно
2                сытный овощной салатик пальчики оближете
3       картофельное пюре и куриные котлеты  вкусная к...
4       вишневая наливка имеет яркий вишневый вкус кот...
                              ...                        
3462    для тех кто любит чечевицу вам сюда очень вкус...
3463    баклажановые фантазии продолжаются предлагаю в...
3464    мое любимое блюдо лазанья но кушать только фар...
3465    прошлым летом варила варенье из одуванчиков по...
3466     и три корочки хлеба  сделал заказ буратино в ...
Name: description, Length: 3467, dtype: object

### Расстояние редактирования

In [30]:
!pip install python-Levenshtein

In [5]:
from Levenshtein import distance as lev

2\. Получите набор уникальных слов `words`, содержащихся в текстах описаний рецептов (воспользуйтесь `word_tokenize` из `nltk`). Сгенерируйте 5 пар случайно выбранных слов и посчитайте между ними расстояние Левенштейна. Выведите на экран результат в следующем виде:

```
d(word1, word2) = x
```

In [6]:
def levenstein(str_1, str_2):
    len_1_n, len_2_m = len(str_1), len(str_2)
    if len_1_n > len_2_m:
        str_1, str_2 = str_2, str_1
        len_1_n, len_2_m = len_2_m, len_1_n

    now_row = range(len_1_n + 1)
    for i in range(1, len_2_m + 1):
        past_row, now_row = now_row, [i] + [0]*len_1_n
        for j in range(1, len_1_n + 1):
            add_value = past_row[j] + 1
            delete_value = now_row[j - 1] + 1
            change_value = past_row[j - 1]
            if str_1[j - 1] != str_2[i - 1]:
                change_value += 1
            now_row[j] = min(add_value, delete_value, change_value)
    return now_row[len_1_n]

In [7]:
text_words = list(set(word_tokenize(" ".join(recipes['description']).replace('\r', ' ').replace('\n', ' '))))
list_indexes = [[random.randint(1, len(text_words)), random.randint(1, len(text_words))] for _ in range(5)]
list_value = [[text_words[indexes[0]], text_words[indexes[1]]] for indexes in list_indexes]

for pare in range(5):
    print(f"d({list_value[pare][0]}, {list_value[pare][1]}) = {levenstein(list_value[pare][0], list_value[pare][1])}")

d(трапезу, устроить) = 7
d(вольные, свклы) = 5
d(бананами, рулетом) = 7
d(представленные, коем) = 13
d(фишка, коллег) = 6


In [8]:
for pare in range(5):
    print(f"d({list_value[pare][0]}, {list_value[pare][1]}) = {lev(list_value[pare][0], list_value[pare][1])}")

d(трапезу, устроить) = 7
d(вольные, свклы) = 5
d(бананами, рулетом) = 7
d(представленные, коем) = 13
d(фишка, коллег) = 6


3\. Напишите функцию, которая принимает на вход 2 текстовые строки `s1` и `s2` и при помощи расстояния Левенштейна определяет, является ли строка `s2` плагиатом `s1`. Функция должна реализовывать следующую логику: для каждого слова `w1` из `s1` проверяет, есть в `s2` хотя бы одно слово `w2`, такое, что расстояние Левенштейна между `w1` и `w2` меньше 2, и считает количество таких слов в `s1` $P$. 

$$ P = \#\{w_1 \in s_1\ | \exists w_2 \in s_2 : d(w_1, w_2) < tol\}$$

$$ L = max(|s1|, |s2|) $$

Здесь $|\cdot|$ - количество слов в строке, $\#A$ - число элементов в множестве $A$, $w \in s$ означает, что слово $w$ содержится в тексте $s$.

Если отношение $P / L$ больше 0.8, то функция должна вернуть True; иначе False.

Продемонстрируйте работу вашей функции на примере описаний двух рецептов с ID 135488 и 851934 (ID рецепта - это число, стоящее в конце url рецепта). Выведите на экран описания этих рецептов и результат работы функции.

In [15]:
def is_plagiarism(s1: str, s2: str) -> bool:
    str_one = word_tokenize(s1)
    str_two = word_tokenize(s2)
    L = max(len(str_one), len(str_two))
    total_lev = []
    for word1 in str_one:
        count_levenstein = []
        for word2 in str_two:
            count_levenstein.append(lev(word1, word2))
        if 0 in count_levenstein or 1 in count_levenstein:
           total_lev.append(1)
        else:
            total_lev.append(0)
    P = sum(total_lev)
    print(P / L)
    return True if (P / L) > 0.8 else False

In [16]:
print('Необходимые нам индексы таблицы с ID = 135488 и 851934: ', end='\n')
[index for index in range(len(list(recipes['url']))) if list(recipes['url'])[index].split('/')[-2] == '135488' or list(recipes['url'])[index].split('/')[-2] == '851934']

Необходимые нам индексы таблицы с ID = 135488 и 851934: 


[958, 1473]

In [17]:
recipes.iloc[958]

url                https://www.povarenok.ru/recipes/show/135488/
name                       Паштет сало-авокадо в хлебных орешках
ingredients    {'Сало': '100 г', 'Соль': '1/3 ч. л.', 'Чеснок...
description    прекрасной закуской к крепким напиткам на фурш...
Name: 958, dtype: object

In [18]:
recipes.iloc[1473]

url                https://www.povarenok.ru/recipes/show/851934/
name                   Паштет из сала и авокадов хлебных орешках
ingredients    {'Сало': '100 г', 'Соль': '1/3 ч. л.', 'Чеснок...
description    замечательной закуской к напиткам на фуршетном...
Name: 1473, dtype: object

In [19]:
recipes.iloc[958]['description'], recipes.iloc[1473]['description']

('прекрасной закуской к крепким напиткам на фуршетном столе станет паштет из сала и авокадо с чесноком который мы спрячем в орешках из бородинского хлеба положив в середину маслину закуска получается необычной оригинальной и ценится особенно мужчинами',
 'замечательной закуской к напиткам на фуршетном столе станет паштет из сала и авокадо с чесноком куда мы добавим орешков из бородинского хлеба положив в середину маслины закуска получается крайне необычной оригинальной и ценится особенно мужчинами')

In [20]:
is_plagiarism(recipes.iloc[958]['description'], recipes.iloc[1473]['description'])

0.8611111111111112


True

  ### Стемминг, лемматизация

4\. На основе набора слов из задания 2 создайте `pd.DataFrame` со столбцами `word`, `stemmed_word` и `normalized_word`. В столбец `stemmed_word` поместите версию слова после проведения процедуры стемминга; в столбец `normalized_word` поместите версию слова после проведения процедуры лемматизации. Столбец `word` укажите в качестве индекса. 

Для стемминга можно воспользоваться `SnowballStemmer` из `nltk`, для лемматизации слов - пакетом `pymorphy2`. Сравните результаты стемминга и лемматизации. Поясните на примере одной из строк получившегося фрейма (в виде текстового комментария), в чем разница между двумя этими подходами.

In [46]:
new_list_value = []
for elems in list_value:
    new_list_value.append(elems[0])
    new_list_value.append(elems[1])
new_list_value

['завтрак',
 'соевый',
 'возвращаюсь',
 'мычание',
 'неприметную',
 'жеребца',
 'пряности',
 'говорит',
 'неразочаровалась',
 'питательное']

In [47]:
morph = pymorphy2.MorphAnalyzer()
stemmer = SnowballStemmer("russian")
stem_lem_df = pd.DataFrame(columns=['word', 'stemmed_word', 'normalized_word'])
stem_lem_df['word'] = new_list_value
stem_lem_df['stemmed_word'] = [stemmer.stem(word) for word in stem_lem_df['word']]
stem_lem_df['normalized_word'] = [morph.parse(word)[0].normalized.word for word in stem_lem_df['word']]
stem_lem_df = stem_lem_df.set_index('word')
stem_lem_df

# Stemmer в целом отбрасывает окончания, суффиксы и другие части слова, оставляя только его корень

stemmed_word   normalized_word
word                                            
завтрак                завтрак           завтрак
соевый                    соев            соевый
возвращаюсь           возвраща      возвращаться
мычание                  мычан           мычание
неприметную          неприметн       неприметный
жеребца                 жеребц           жеребец
пряности               пряност          пряность
говорит                  говор          говорить
неразочаровалась  неразочарова  неразочароваться
питательное          питательн       питательный

5\. Добавьте в таблицу `recipes` столбец `description_no_stopwords`, в котором содержится текст описания рецепта после удаления из него стоп-слов. Посчитайте и выведите на экран долю стоп-слов среди общего количества слов. Сравните топ-10 самых часто употребляемых слов до и после удаления стоп-слов.

In [22]:
nltk.download('stopwords')
from nltk.corpus import stopwords

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\admin\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\stopwords.zip.


In [48]:
stops = stopwords.words('russian')
pattern = r'\b(?:{})\b'.format('|'.join(stops))
recipes['description_no_stopwords'] = recipes['description'].str.replace(pattern, '')

C:\Users\admin\AppData\Local\Temp/ipykernel_5244/3367714406.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  recipes['description_no_stopwords'] = recipes['description'].str.replace(pattern, '')


In [54]:
word_counts_before = Counter(word_tokenize(' '.join(recipes['description'])))
count_before, count_stops = sum(word_counts_before.values()), sum([word_counts_before[x] for x in stops])
count_before, count_stops, count_stops / count_before

(102978, 33223, 0.322622307677368)

In [56]:
word_counts_after = Counter(word_tokenize(' '.join(recipes['description_no_stopwords'])))
word_count_before = word_counts_before.most_common(10)
word_count_after = word_counts_after.most_common(10)
sum(word_counts_after.values()), count_before - count_stops

(69755, 69755)

In [57]:
word_count_before

[('и', 5054),
 ('в', 2584),
 ('с', 1934),
 ('на', 1655),
 ('очень', 1607),
 ('не', 1517),
 ('из', 1006),
 ('я', 979),
 ('рецепт', 869),
 ('а', 863)]

In [58]:
word_count_after

[('очень', 1607),
 ('рецепт', 869),
 ('это', 734),
 ('блюдо', 524),
 ('вкусный', 461),
 ('просто', 436),
 ('вкусно', 375),
 ('приготовить', 344),
 ('вкус', 324),
 ('салат', 313)]

### Векторное представление текста

6\. Выберите случайным образом 5 рецептов из набора данных, в названии которых есть слово "оладьи" (без учета регистра). Представьте описание каждого рецепта в виде числового вектора при помощи `TfidfVectorizer`. На основе полученных векторов создайте `pd.DataFrame`, в котором названия колонок соответствуют словам из словаря объекта-векторизатора. 

Примечание: обратите внимание на порядок слов при создании колонок.

In [59]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [215]:
olad = recipes[recipes['name'].str.contains('оладьи')]
sample = olad.sample(n=5)
sample

url  \
3117  https://www.povarenok.ru/recipes/show/122540/   
1139    https://www.povarenok.ru/recipes/show/6443/   
464   https://www.povarenok.ru/recipes/show/133050/   
838    https://www.povarenok.ru/recipes/show/20041/   
1601    https://www.povarenok.ru/recipes/show/1239/   

                                      name  \
3117  Картофельные оладьи с соусом "Весна"   
1139                        Куриные оладьи   
464                        Яблочные оладьи   
838                       Тыквенные оладьи   
1601          Картофельные оладьи с семгой   

                                            ingredients  \
3117  {'Картофель': '600 г', 'Лук репчатый': '1 шт',...   
1139  {'Грудка куриная': None, 'Лук репчатый': '2 шт...   
464   {'Яблоко': '2 шт', 'Яйцо куриное': '1 шт', 'Му...   
838   {'Тыква': '500 г', 'Яйцо куриное': '2 шт', 'Кр...   
1601  {'Картофель': '3 шт', 'Семга': '200 г', 'Крахм...   

                                            description  \
3117  оладьи могут быть разные эти  картофельные с р...   
1139  на скорую руку за фото к рецепту спасибо людми...   
464   очень простое бюджетное блюдо особенно понрави...   
838        нежные вкусные оладушки  мужу очень нравятся   
1601  вкусные и сытные оладушки с явным вкусом жарен...   

                               description_no_stopwords  
3117  оладьи могут  разные   картофельные  рыбой  до...  
1139   скорую руку  фото  рецепту спасибо людмиле сурик  
464   очень простое бюджетное блюдо особенно понрави...  
838        нежные вкусные оладушки  мужу очень нравятся  
1601  вкусные  сытные оладушки  явным вкусом жареной...

In [216]:
def get_dataframe_vect(sent_words):
    cv = TfidfVectorizer()
    cv.fit(sent_words)
    return pd.DataFrame([{elem: cv.transform(sent_words).toarray()[0][cv.vocabulary_[elem]] for elem in cv.vocabulary_}])

get_dataframe_vect(sent_tokenize(sample['description'].values[0]))

оладьи     могут      быть    разные       эти  картофельные     рыбой  \
0  0.158114  0.158114  0.158114  0.158114  0.158114      0.158114  0.158114   

   дополнены  сметанным    соусом  ...  временем     годом  покупать  \
0   0.158114   0.158114  0.158114  ...  0.158114  0.158114  0.158114   

         мы        их  начинали     марта    месяца        на  праздник  
0  0.158114  0.158114  0.158114  0.158114  0.158114  0.158114  0.158114  

[1 rows x 34 columns]

In [217]:
get_dataframe_vect(sent_tokenize(sample['description'].values[1]))

на    скорую      руку        за      фото   рецепту   спасибо  \
0  0.333333  0.333333  0.333333  0.333333  0.333333  0.333333  0.333333   

    людмиле     сурик  
0  0.333333  0.333333

In [218]:
get_dataframe_vect(sent_tokenize(sample['description'].values[2]))

очень   простое  бюджетное     блюдо  особенно  понравится       тем  \
0  0.301511  0.301511   0.301511  0.301511  0.301511    0.301511  0.301511   

       кому    нельзя  молочные  продукты  
0  0.301511  0.301511  0.301511  0.301511

In [219]:
get_dataframe_vect(sent_tokenize(sample['description'].values[3]))

нежные   вкусные  оладушки      мужу     очень  нравятся
0  0.408248  0.408248  0.408248  0.408248  0.408248  0.408248

In [220]:
get_dataframe_vect(sent_tokenize(sample['description'].values[4]))

вкусные    сытные  оладушки     явным    вкусом   жареной      смги
0  0.377964  0.377964  0.377964  0.377964  0.377964  0.377964  0.377964

7\. Вычислите близость между каждой парой рецептов, выбранных в задании 6, используя косинусное расстояние (можно воспользоваться функциями из любого пакета: `scipy`, `scikit-learn` или реализовать функцию самому). Результаты оформите в виде таблицы `pd.DataFrame`. В качестве названий строк и столбцов используйте названия рецептов.

Примечание: обратите внимание, что $d_{cosine}(x, x) = 0$

In [221]:
sample_desc = [sample.iloc[x]['description'] for x in range(5)]
vecter = CountVectorizer().fit_transform(sample_desc).toarray()
csim = cosine_similarity(vecter)
csim

array([[1.        , 0.05270463, 0.        , 0.        , 0.        ],
       [0.05270463, 1.        , 0.        , 0.        , 0.        ],
       [0.        , 0.        , 1.        , 0.12309149, 0.        ],
       [0.        , 0.        , 0.12309149, 1.        , 0.3086067 ],
       [0.        , 0.        , 0.        , 0.3086067 , 1.        ]])

In [222]:
df_olad = pd.DataFrame(csim, columns=sample['name'].values, index=sample['name'].values)
for i in range(5):
    df_olad.iloc[i][i] = 0
df_olad

Картофельные оладьи с соусом "Весна"  \
Картофельные оладьи с соусом "Весна"                              0.000000   
Куриные оладьи                                                    0.052705   
Яблочные оладьи                                                   0.000000   
Тыквенные оладьи                                                  0.000000   
Картофельные оладьи с семгой                                      0.000000   

                                      Куриные оладьи  Яблочные оладьи  \
Картофельные оладьи с соусом "Весна"        0.052705         0.000000   
Куриные оладьи                              0.000000         0.000000   
Яблочные оладьи                             0.000000         0.000000   
Тыквенные оладьи                            0.000000         0.123091   
Картофельные оладьи с семгой                0.000000         0.000000   

                                      Тыквенные оладьи  \
Картофельные оладьи с соусом "Весна"          0.000000   
Куриные оладьи                                0.000000   
Яблочные оладьи                               0.123091   
Тыквенные оладьи                              0.000000   
Картофельные оладьи с семгой                  0.308607   

                                      Картофельные оладьи с семгой  
Картофельные оладьи с соусом "Весна"                      0.000000  
Куриные оладьи                                            0.000000  
Яблочные оладьи                                           0.000000  
Тыквенные оладьи                                          0.308607  
Картофельные оладьи с семгой                              0.000000

8\. Напишите функцию, которая принимает на вход `pd.DataFrame`, полученный в задании 7, и возвращает в виде кортежа названия двух различных рецептов, которые являются наиболее похожими. Прокомментируйте результат (в виде текстового комментария). Для объяснения результата сравните слова в описаниях двух этих отзывов.

In [231]:
def find_closest(sim_df: pd.DataFrame) -> tuple:
    print(sim_df[sim_df == sim_df.max().max()].stack().values[0])
    return sim_df[sim_df == sim_df.max().max()].stack().index.tolist()[0]
find_closest(df_olad)

0.3086066999241838


('Тыквенные оладьи', 'Картофельные оладьи с семгой')

In [232]:
sample[sample['name']==find_closest(df_olad)[0]]['description'].values[0]

0.3086066999241838


'нежные вкусные оладушки  мужу очень нравятся'

In [233]:
sample[sample['name']==find_closest(df_olad)[1]]['description'].values[0]

0.3086066999241838


'вкусные и сытные оладушки с явным вкусом жареной смги'

##### Вывод о схожести описания 2 рецептов
Здесь видно, что есть полностью совпавшее словосочетание "вкусные оладушки", которое составляет 1/3 слов в первом и втором предложениях, а результат работы функции (синусное расстояние, коэффициент) составляет 0.308606 (30.86%).